src/hooks/context/useSendersBalancesContext.ts

```diff
@@ -0,0 +1,98 @@
+
+import { Dispatch, useEffect } from 'react';
+import BN from 'bn.js';
+import { AccountsActionType, AccountState } from '../../types/accountTypes';
+import { useApiCallsContext } from '../../contexts/ApiCallsContextProvider';
+import { useSubscriptionsContext } from '../../contexts/SubscriptionsContextProvider';
+import { useMountedState } from '../react/useMountedState';
+import { BLOCK_DURATION, TIMER_DURATION } from '../../constants';
+
+const increaseBlock = (current: BN) => current.add(BLOCK_DURATION);
+const getBlockAsBN = (block: string) => new BN(block);
+
+const convertToBNAndIncrease = (block: string) => increaseBlock(getBlockAsBN(block));
+
+export default function useSendersBalancesContext(
+  accountState: AccountState,
+  dispatchAccount: Dispatch<AccountsActionType>
+) {
+  const {
+    sourceSubscriptions: { bestBlock: currentSourceBlock },
+    targetSubscriptions: { bestBlock: currentTargetBlock }
+  } = useSubscriptionsContext();
+
+  const [nextSourceBlockNumber, setNextSourceBlockNumber] = useMountedState<BN>(
+    convertToBNAndIncrease(currentSourceBlock)
+  );
+  const [nextTargetBlockNumber, setNextTargetBlockNumber] = useMountedState<BN>(
+    convertToBNAndIncrease(currentTargetBlock)
+  );
+
+  const [timerId, setTimerId] = useMountedState<NodeJS.Timeout | null>(null);
+
+  const { updateSenderAccountsInformation } = useApiCallsContext();
+
+  useEffect(() => {
+    const currentBNSource = getBlockAsBN(currentSourceBlock);
+    const currentBNTarget = getBlockAsBN(currentTargetBlock);
+    if (currentBNSource.gte(nextSourceBlockNumber) && currentBNTarget.gte(nextTargetBlockNumber)) {
+      updateSenderAccountsInformation(dispatchAccount);
+      setNextSourceBlockNumber(increaseBlock(currentBNSource));
+      setNextTargetBlockNumber(increaseBlock(currentBNTarget));
+      if (timerId) {
+        clearTimeout(timerId);
+        setTimerId(null);
+      }
+    }
+  }, [
+    currentSourceBlock,
+    currentTargetBlock,
+    dispatchAccount,
+    nextSourceBlockNumber,
+    nextTargetBlockNumber,
+    setNextSourceBlockNumber,
+    setNextTargetBlockNumber,
+    setTimerId,
+    timerId,
+    updateSenderAccountsInformation
+  ]);
+
+  useEffect(() => {
+    if (!timerId) {
+      const id = setTimeout(() => {
+        updateSenderAccountsInformation(dispatchAccount);
+        const increasedSourceBlock = increaseBlock(nextSourceBlockNumber);
+        const increasedTargetBlock = increaseBlock(nextTargetBlockNumber);
+        setNextSourceBlockNumber(increasedSourceBlock);
+        setNextTargetBlockNumber(increasedTargetBlock);
+        clearTimeout(timerId!);
+        setTimerId(null);
+      }, TIMER_DURATION);
+      setTimerId(id);
+    }
+  }, [
+    dispatchAccount,
+    nextSourceBlockNumber,
+    nextTargetBlockNumber,
+    setNextSourceBlockNumber,
+    setNextTargetBlockNumber,
+    setTimerId,
+    timerId,
+    updateSenderAccountsInformation
+  ]);
+}
```

Comment:
```suggestion  
  const updateAccounts = useCallback(() => {
        updateSenderAccountsInformation(dispatchAccount);
        setNextSourceBlockNumber(convertToBNAndIncrease(currentSourceBlock));
        setNextTargetBlockNumber(convertToBNAndIncrease(currentTargetBlock));
        // clear timer id if any
        if (timerId) {
          clearTimeout(timerId);
        }
        // set new timer 
        const newTimerId = setTimeout(updateAccounts, TIMER_DURATION);
        setTimerId(newTimerId);
  }, [
    dispatchAccount,
    currentSourceBlock,
    currentTargetBlock,
    setNextSourceBlockNumber,
    setNextTargetBlockNumber,
    setTimerId,
    timerId,
    updateSenderAccountsInformation
  ]);
  
  const blocksReached = getBlockAsBN(currentSourceBlock).gte(nextSourceBlockNumber)
    && getBlockAsBN(currentTargetBlock).gte(nextTargetBlockNumber);
    
  useEffect(() => {
    if (blocksReached) {
       updateAccounts();
    }
    // we need to make sure to cancel the timeout, this is enough to prevent setState after render,
    // so then we don't have to use `useMountedState` but rather can use the regular one.
    return () => {
      timerId && clearTimeout(timerId);
    };
  }, [blocksReached, updateAccounts, timerId]);
}
```

This is a bit shorter and imho much cleaner. By having only single state object for both `nextSource` and `nextTarget` block we could reduce the code (especially the deps) even more.

Reply:
The code looks very good, but it's not so accurate as my original code. i'm running this version and i'm logging every update with a timestamp , and i have verified that the updates are not happening with the expected frequency. In this case the updates should be every 24 seconds if there is no big delay ( block duration = 6 seconds * 4 ).

Also i did the test to reduce the timer to 10 seconds looking forward to get the update every 10 seconds and the update happens every 22 / 23 seconds.

Reply:
Aren't you curious WHY it does not work? Like if you agree that the code is "good" but you dismiss it because you can't figure out why it doesn't work as expected, then there is no point in merging this code.
Maybe try removing the cleanup function from `useEffect`?

The point of leaving code suggestions is always to challenge the existing solution and encourage you to go a bit of extra mile to figure out a better solution (cause you seem to agree that the current one is sub optimal). If you're not interested, I don't see much point of doing further reviews.

Reply:
i was just saying pointing out that the code is not working as expected and i have never said that i'm not interested , actually i consider this as kind of work in progress to be honest, but i wanted to provide the feedback of what it is not working.

> Maybe try removing the cleanup function from useEffect?

I will try this out.

Reply:
Your code works great removing the cleanup function from useEffect as you proposed. My understanding is that if the timer is running and the component gets unmounted, the clean up will happen anyway because is part of the `updateAccounts` , so we would be covered on that. Anyway, the `useMountedState` will be still needed .